app/data_sources/azuredevops.py

```diff
@@ -0,0 +1,120 @@
+import datetime
+import logging
+from typing import Dict, List
+import requests
+import base64
+import urllib.parse
+from dataclasses import dataclass
+
+from azure.devops.connection import Connection
+from msrest.authentication import BasicAuthentication
+
+from data_source_api.base_data_source import BaseDataSource, ConfigField, HTMLInputType
+from data_source_api.basic_document import DocumentType, BasicDocument
+from data_source_api.exception import InvalidDataSourceConfig
+from index_queue import IndexQueue
+from pydantic import BaseModel
+from parsers.html import html_to_text
+from data_source_api.utils import parse_with_workers
+
+logger = logging.getLogger(__name__)
+
+@dataclass
+class DevOpsConfig():
+    organization_url: str
+    access_token: str
+    project_name: str
+    query_id: str
+
+    def __post_init__(self):
+        self.query_id = self.query_id.strip()
+        self.access_token = self.access_token.strip()
+        self.project_name = self.project_name.strip()
+        self.organization_url = self.organization_url.strip()
+
+class AzuredevopsDataSource(BaseDataSource):
+    @staticmethod
+    def get_config_fields() -> List[ConfigField]:
+        return [
+            ConfigField(label="AzureDevOps organization URL", placeholder="https://dev.azure.com/org", name="organization_url"),
+            ConfigField(label="Personal Access Token", name="access_token", type=HTMLInputType.PASSWORD),
+            ConfigField(label="Project Name", name="project_name"),
+            ConfigField(label="Query ID", name="query_id"),
+        ]
+    
+    @staticmethod
+    def validate_config(config: Dict) -> None:
+        try:
+            devops_config = DevOpsConfig(**config)
+            credentials = BasicAuthentication('', devops_config.access_token)
+            connection = Connection(base_url=devops_config.organization_url, creds=credentials)
+            core_client = connection.clients.get_core_client()
+            core_client.get_projects()
+        except Exception as e:
+            raise InvalidDataSourceConfig from e
+        
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._config['devops_config'] = DevOpsConfig(**self._config)
```

Comment:
@Itaykal I was attempting to save it to `self._config.devops_config` after a discussion I had with @Roey7, but it was giving an attribute error. So I went this route. If it would be preferred I can change this to be self._devops_config.

Reply:
I think it would be more redable as an attribute and not a dict item